cirq-google/cirq_google/workflow/qubit_placement.py

```diff
@@ -107,6 +108,72 @@ def default_topo_node_to_qubit(node: Any) -> cirq.Qid:
         return cirq.LineQubit(node)
 
 
+class HardcodedQubitPlacer(QubitPlacer):
+    def __init__(
+        self,
+        mapping: Dict[cirq.NamedTopology, Dict[Any, cirq.Qid]],
+        topo_node_to_qubit_func: Callable[[Hashable], cirq.Qid] = default_topo_node_to_qubit,
+    ):
+        self.mapping = mapping
+        self.topo_node_to_qubit_func = topo_node_to_qubit_func
+
+    def place_circuit(
+        self,
+        circuit: cirq.AbstractCircuit,
+        problem_topology: NamedTopology,
+        shared_rt_info: 'cg.SharedRuntimeInfo',
+        rs: np.random.RandomState,
+    ) -> Tuple[cirq.FrozenCircuit, Dict[Any, cirq.Qid]]:
+        try:
+            nt_mapping = self.mapping[problem_topology]
+        except KeyError as e:
+            raise CouldNotPlaceError(str(e))
+
+        circuit_mapping = {
+            self.topo_node_to_qubit_func(nt_node): gridq for nt_node, gridq in nt_mapping.items()
+        }
+
+        circuit = circuit.unfreeze().transform_qubits(circuit_mapping).freeze()
+        return circuit, circuit_mapping
+
+    def __repr__(self) -> str:
+        return f'cirq_google.HardcodedQubitPlacer(mapping={_compat.proper_repr(self.mapping)})'
+
+    @classmethod
+    def _json_namespace_(cls) -> str:
+        return 'cirq.google'
+
+    def _json_dict_(self):
+        d = obj_to_dict_helper(self, attribute_names=[])
+
+        # Nested dict: turn both levels to list(key_value_pair)
+        mapping = {topo: list(placement.items()) for topo, placement in self.mapping.items()}
+        mapping = list(mapping.items())
+        d['mapping'] = mapping
+        return d
+
+    @classmethod
+    def _from_json_dict_(cls, **kwargs):
+        # From nested list(key_value_pair) to dictionary
+        mapping: Dict[cirq.NamedTopology, Dict[Any, 'cirq.Qid']] = {}
+        for topo, placement_kvs in kwargs['mapping']:
+            placement: Dict[Hashable, 'cirq.Qid'] = {}
+            for k, v in placement_kvs:
+                if isinstance(k, list):
+                    k = tuple(k)
+                placement[k] = v
+            mapping[topo] = placement
+
+        return cls(mapping=mapping)
+
+    def __eq__(self, other):
+        if not isinstance(other, HardcodedQubitPlacer):
+            # coverage: ignore
+            return False
+
+        return self.mapping == other.mapping
```

Comment:
Consider using `@cirq.value_equality` so that other nice helpers, like __ne__ and __hash__, are automatically added?

Reply:
we get `__ne__` by defining `__eq__` (added a test). I don't think hash is appropriate since it's backed by a non-hashable dictionary of dictionaries. Nor do I really see the use for hashing this. We don't need comparison operators either.